test/subscription.js

```diff
@@ -508,3 +513,316 @@ test('subscription connection is closed if context function throws', t => {
+    Query: {
+      notifications: () => notifications
+    },
+    Mutation: {
+      addNotification: async (_, { message }) => {
+        const id = idCount++
+        const notification = {
+          id,
+          message
+        }
+        notifications.push(notification)
+        await emitter.emit({
+          topic: 'NOTIFICATION_ADDED',
+          payload: {
+            notificationAdded: notification
+          }
+        })
+
+        return notification
+      }
+    },
+    Subscription: {
+      notificationAdded: {
+        subscribe: (root, args, { pubsub, topic }) => pubsub.subscribe(topic)
+      }
+    }
+  }
+
+  app.register(GQL, {
+    schema,
+    resolvers,
+    subscription: {
+      emitter,
+      context: async () => {
+        await wait(200)
+        return { topic: 'NOTIFICATION_ADDED' }
+      }
+    }
+  })
+
+  app.listen(0, err => {
+    t.error(err)
+
+    const ws = new WebSocket('ws://localhost:' + (app.server.address()).port + '/graphql', 'graphql-ws')
+    const client = WebSocket.createWebSocketStream(ws, { encoding: 'utf8', objectMode: true })
+    t.tearDown(client.destroy.bind(client))
+    client.setEncoding('utf8')
+
+    client.write(JSON.stringify({
+      type: 'connection_init'
+    }))
+
+    client.write(JSON.stringify({
+      id: 1,
+      type: 'start',
+      payload: {
+        query: `
+          subscription {
+            notificationAdded {
+              id
+              message
+            }
+          }
+        `
+      }
+    }))
+
+    client.write(JSON.stringify({
+      id: 2,
+      type: 'start',
+      payload: {
+        query: `
+          subscription {
+            notificationAdded {
+              id
+              message
+            }
+          }
+        `
+      }
+    }))
+
+    client.write(JSON.stringify({
+      id: 2,
+      type: 'stop'
+    }))
+
+    client.on('data', chunk => {
+      const data = JSON.parse(chunk)
+
+      if (data.id === 1 && data.type === 'data') {
+        t.equal(chunk, JSON.stringify({
+          type: 'data',
+          id: 1,
+          payload: {
+            data: {
+              notificationAdded: {
+                id: '1',
+                message: 'Hello World'
+              }
+            }
+          }
+        }))
+
+        client.end()
+        t.end()
+      } else if (data.id === 2 && data.type === 'complete') {
+        sendTestQuery()
+      }
+    })
+  })
+})
+
+test('subscription connection is closed if async context function throws', t => {
+  const app = Fastify()
+  t.tearDown(() => app.close())
+
+  const schema = `
+    type Query {
+      add(x: Int, y: Int): Int
+    }
+  `
+
+  const resolvers = {
+    Query: {
+      add: (parent, { x, y }) => x + y
+    }
+  }
+
+  const wait = (amount = 0) => new Promise(resolve => setTimeout(resolve, amount))
+
+  app.register(GQL, {
+    schema,
+    resolvers,
+    subscription: {
+      context: async function () {
+        await wait(200)
```

Comment:
Why is this timeout needed?

Reply:
this timeout triggers the race condition of the socket not being created because of the context creation await, this simulates the reason this PR was created

Reply:
Then you can use the following instead:

```js
const immediate = promisify(setImmediate)

// in the test
await immediate()
```

Reply:
with setTimeout(() => {}, 0), which is what setImmediate does, the race condition was not happening, the number should be high for it to happen.

Reply:
the problem was not happening just because it was async, was because of the delay intrinsic the async, which is we are simulating with the delay 200, you can test this problem yourself removing the code that solves this and keeping the tests.

Reply:
Can we use the fake timers instead then? setTimeout introduce flakyness.

Reply:
are they being installed/used already somewhere in the code?

Reply:
https://github.com/mcollina/fastify-gql/blob/1335259f51710521b3d81cfb01e78d84d241577c/test/gateway/pollingInterval.js#L5 use the fake timers.

Reply:
I was checking the timers and they don't do what the test needs need to do, we need to actually make the socket wait, and the number needs to be 50ms+, the flakyness doesn't matter